prometheus/benchmark-tests.txt

```diff
@@ -0,0 +1,44 @@
+PrometheusReporterBenchmark.reportCounterBenchmark:·gc.churn.G1_Old_Gen.norm                      thrpt   10     0.034 ±     0.025    B/op
+PrometheusReporterBenchmark.reportCounterBenchmark:·gc.count                                      thrpt   10   507.000              counts
+PrometheusReporterBenchmark.reportCounterBenchmark:·gc.time                                       thrpt   10   498.000                  ms
+PrometheusReporterBenchmark.reportGaugeBenchmark                                                  thrpt   10   242.631 ±    25.759  ops/ms
+PrometheusReporterBenchmark.reportGaugeBenchmark:·gc.alloc.rate                                   thrpt   10  1407.404 ±   150.312  MB/sec
+PrometheusReporterBenchmark.reportGaugeBenchmark:·gc.alloc.rate.norm                              thrpt   10  6388.006 ±     6.374    B/op
+PrometheusReporterBenchmark.reportGaugeBenchmark:·gc.churn.G1_Eden_Space                          thrpt   10  1406.450 ±   157.275  MB/sec
+PrometheusReporterBenchmark.reportGaugeBenchmark:·gc.churn.G1_Eden_Space.norm                     thrpt   10  6382.387 ±    92.561    B/op
+PrometheusReporterBenchmark.reportGaugeBenchmark:·gc.churn.G1_Old_Gen                             thrpt   10     0.008 ±     0.005  MB/sec
+PrometheusReporterBenchmark.reportGaugeBenchmark:·gc.churn.G1_Old_Gen.norm                        thrpt   10     0.036 ±     0.022    B/op
+PrometheusReporterBenchmark.reportGaugeBenchmark:·gc.count                                        thrpt   10   486.000              counts
+PrometheusReporterBenchmark.reportGaugeBenchmark:·gc.time                                         thrpt   10   468.000                  ms
+PrometheusReporterBenchmark.reportHistogramDurationSamplesBenchmark                               thrpt   10     0.693 ±     0.102  ops/ms
+PrometheusReporterBenchmark.reportHistogramDurationSamplesBenchmark:·gc.alloc.rate                thrpt   10     4.125 ±     0.609  MB/sec
+PrometheusReporterBenchmark.reportHistogramDurationSamplesBenchmark:·gc.alloc.rate.norm           thrpt   10  6558.180 ±    95.845    B/op
+PrometheusReporterBenchmark.reportHistogramDurationSamplesBenchmark:·gc.churn.G1_Eden_Space       thrpt   10     5.749 ±    18.325  MB/sec
+PrometheusReporterBenchmark.reportHistogramDurationSamplesBenchmark:·gc.churn.G1_Eden_Space.norm  thrpt   10  9188.893 ± 29287.662    B/op
+PrometheusReporterBenchmark.reportHistogramDurationSamplesBenchmark:·gc.count                     thrpt   10     2.000              counts
+PrometheusReporterBenchmark.reportHistogramDurationSamplesBenchmark:·gc.time                      thrpt   10     4.000                  ms
+PrometheusReporterBenchmark.reportHistogramValueSamplesBenchmark                                  thrpt   10     0.777 ±     0.120  ops/ms
+PrometheusReporterBenchmark.reportHistogramValueSamplesBenchmark:·gc.alloc.rate                   thrpt   10     4.355 ±     0.653  MB/sec
```

Comment:
Something is off in this benchmark: 
- It has substantially less alloc rate than counters (which is unlikely)
- It has only 2 GCs occurring

Reply:
i think it makes sense: the benchmark test has a long counter which goes up to Long.MAX_VALUE and with current implementation of histogram the tests keeps busy iterating over this huge long value and as a result there are not many allocations and not that many GCs happening. Need to fix this reportHistogram* methods

Reply:
> the tests keeps busy iterating over this huge long value and as a result there are not many allocations and not that many GCs happening

Not sure I get why we're iterating up to Long.MAX_VALUE in the test; are we testing the right thing here?

Reply:
Sorry for the long reply. Need to configure notifications.
my reasoning there is:
1. we use long type as an input param => the value can potentially go up to the Long.MAX_VALUE (even though highly unlikely);
2. the value resets after each report interval, which is configurable on the client side. From what I found in the usages the report interval varies from 1s to few minutes;
3. Other implementations do not depend on the value range;
4. I couldn't reason for any other particular value as a threshold, so why not go up to the max;
5. It's a good indicator we need to improve it :)

Reply:
Thanks for explanation! 

> Other implementations do not depend on the value range;

I didn't realize that this one did haha; definitely agreed that we should figure out how to not do that here. Will look more into that part of the implementation to develop a more informed opinion.